pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -1533,40 +1591,49 @@ func (gce *GCECloud) DeleteRoute(clusterName string, route *cloudprovider.Route)
 }
 
 func (gce *GCECloud) GetZone() (cloudprovider.Zone, error) {
-	region, err := getGceRegion(gce.zone)
-	if err != nil {
-		return cloudprovider.Zone{}, err
-	}
 	return cloudprovider.Zone{
-		FailureDomain: gce.zone,
-		Region:        region,
+		FailureDomain: gce.instanceZone,
+		Region:        gce.region,
 	}, nil
 }
 
-func (gce *GCECloud) CreateDisk(name string, sizeGb int64) error {
+// Create a new Persistent Disk, with the specified name & size.
+// If zone is specified (!= ""), the disk will be created in that zone.
+// If the zone is not specified (""), the disk will be created in the zone in which we are currently running.
+// Currently, this will be the zone where the kube-controller-manager is running.
+// TODO: We should support auto-provisioning volumes in multiple/specified zones
+func (gce *GCECloud) CreateDisk(name string, zone string, sizeGb int64) error {
 	diskToCreate := &compute.Disk{
 		Name:   name,
 		SizeGb: sizeGb,
 	}
-	createOp, err := gce.service.Disks.Insert(gce.projectID, gce.zone, diskToCreate).Do()
+	if zone == "" {
+		zone = gce.instanceZone
+	}
+	createOp, err := gce.service.Disks.Insert(gce.projectID, zone, diskToCreate).Do()
 	if err != nil {
 		return err
 	}
 
-	return gce.waitForZoneOp(createOp)
+	return gce.waitForZoneOp(createOp, zone)
 }
 
 func (gce *GCECloud) DeleteDisk(diskToDelete string) error {
-	deleteOp, err := gce.service.Disks.Delete(gce.projectID, gce.zone, diskToDelete).Do()
+	disk, err := gce.resolveDisk(diskToDelete)
 	if err != nil {
 		return err
 	}
 
-	return gce.waitForZoneOp(deleteOp)
+	deleteOp, err := gce.service.Disks.Delete(gce.projectID, disk.Zone, disk.Name).Do()
+	if err != nil {
+		return err
+	}
+
+	return gce.waitForZoneOp(deleteOp, disk.Zone)
 }
 
 func (gce *GCECloud) AttachDisk(diskName, instanceID string, readOnly bool) error {
-	disk, err := gce.getDisk(diskName)
+	disk, err := gce.resolveDisk(diskName)
```

Comment:
Here you can get the zone from the instance, as it only makes sense to attach them if they're in the same zone.